90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -1216,50 +1216,93 @@ read_kcl_prop() {
+    zerror "fs is undefined"
     return 1
   fi
-  zdebug "zfsbe_mnt set to ${zfsbe_mnt}"
+  zdebug "fs set to ${fs}"
 
-  args_file="${BASE}/${zfsbe_fs}/cmdline"
+  args_file="${BASE}/${fs}/cmdline"
 
-  if zfsbe_args="$( read_kcl_prop "${zfsbe_fs}" )" && [ -n "${zfsbe_args}" ]; then
+  if args="$( read_kcl_prop "${fs}" )" && [ -n "${args}" ]; then
     zdebug "using org.zfsbootmenu:commandline"
-    echo "${zfsbe_args}" > "${args_file}"
-    return
+    echo "${args}" > "${args_file}"
+    return 0
   fi
 
-  if [ -n "${zfsbe_mnt}" ] && [ -r "${zfsbe_mnt}/etc/default/zfsbootmenu" ]; then
-    zdebug "using ${zfsbe_mnt}/etc/default/zfsbootmenu"
-    head -1 "${zfsbe_mnt}/etc/default/zfsbootmenu" | tr -d '\n' > "${args_file}"
-    return
+  # Mount R/O to check for config files
+  if ! mnt="$( mount_zfs "${fs}" )"; then
+    zerror "unable to mount ${fs}"
+    return 1
   fi
 
-  if [ -n "${zfsbe_mnt}" ] && [ -r "${zfsbe_mnt}/etc/default/grub" ]; then
-    zdebug "using ${zfsbe_mnt}/etc/default/grub"
+  if [ -r "${mnt}/etc/default/zfsbootmenu" ]; then
+    zdebug "using ${mnt}/etc/default/zfsbootmenu"
+    head -1 "${mnt}/etc/default/zfsbootmenu" | tr -d '\n' > "${args_file}"
+    deprecated="/etc/default/zfsbootmenu"
+  elif [ -r "${mnt}/etc/default/grub" ]; then
+    zdebug "using ${mnt}/etc/default/grub"
     echo "$(
       # shellcheck disable=SC1090,SC1091
-      . "${zfsbe_mnt}/etc/default/grub" ;
+      . "${mnt}/etc/default/grub" ;
       echo "${GRUB_CMDLINE_LINUX_DEFAULT}"
     )" > "${args_file}"
-    return
+    deprecated="/etc/default/grub"
   fi
+
+  # Always unmount, pool must be writable to perform migration
+  umount "${mnt}" || return 1
+
+  if [ "${deprecated}" = "/etc/default/zfsbootmenu" ]; then
+    # Need an R/W mount to remove deprecated config
+    need_rw="yes"
+  elif [ -z "${deprecated}" ] || [ -n "${zbm_ignore_kcl_deprecation}" ]; then
+    # Nothing is deprecated, so there is nothing to do
+    return 0
+  fi
+
+  # It is not an error if user declines automatic migration
+  if ! color=green delay=60 prompt="Will attempt migration in %0.2d seconds" \
+      timed_prompt "Using KCL from ${deprecated} on ${fs}" \
+      "This behavior is DEPRECATED and will be removed soon" "" \
+      "KCL should be migrated to an org.zfsbootmenu:commandline property" "" \
+      "[RETURN] to migrate" "[ESCAPE] to ignore "; then
+    # Suppress repeated messages
+    export zbm_ignore_kcl_deprecation=1
+    echo 'export zbm_ignore_kcl_deprecation="1"' >> /etc/zfsbootmenu.conf
+    return 0
+  fi
+
+  zdebug "migrating ${deprecated} to org.zfsbootmenu:commandline"
+
+  # Pool must be writable to set property and remove config
+  set_rw_pool "${fs%%/*}" || return 1
+  CLEAR_SCREEN=1 load_key "${fs}"
+
+  if ! mnt="$( allow_rw="${need_rw}" mount_zfs "${fs}" )"; then
+    zerror "unable to mount ${fs}"
+    return 1
+  fi
+
+  read -r args < "${args_file}"
+
+  if ! zfs set org.zfsbootmenu:commandline="${args}" "${fs}" ; then
```

Comment:
It might be good to capture the output of this `zfs set` so that it can be logged through either `zdebug` or `zerror` if it fails for some reason.

Reply:
Done